d2ir/compile.go

```diff
@@ -0,0 +1,251 @@
+package d2ir
+
+import (
+	"oss.terrastruct.com/d2/d2ast"
+	"oss.terrastruct.com/d2/d2parser"
+)
+
+type compiler struct {
+	err d2parser.ParseError
+}
+
+func (c *compiler) errorf(n d2ast.Node, f string, v ...interface{}) {
+	c.err.Errors = append(c.err.Errors, d2parser.Errorf(n, f, v...).(d2ast.Error))
+}
+
+func Compile(ast *d2ast.Map) (*Map, error) {
+	c := &compiler{}
+	m := &Map{}
+	m.initRoot()
+	m.parent.(*Field).References[0].Context.Scope = ast
+	c.compileMap(m, ast)
+	c.compileScenarios(m)
+	c.compileSteps(m)
+	if !c.err.Empty() {
+		return nil, c.err
+	}
+	return m, nil
+}
+
+func (c *compiler) compileScenarios(m *Map) {
+	scenariosf := m.GetField("scenarios")
+	if scenariosf == nil {
+		return
+	}
+	scenarios := scenariosf.Map()
+	if scenarios == nil {
+		return
+	}
+
+	for _, sf := range scenarios.Fields {
+		if sf.Map() == nil {
+			continue
```

Comment:
shouldn't we just make a copy anyway? even if no changes

Reply:
Neither of these checks will work anyway because like you pointed out Map() is never nil on `*Field` lol.

I just removed them.

Reply:
Nvm given my previous comment. So why should we make a copy anyway?

Reply:
just allows progressively building up. ppl wanna test layers, they make 3, they're empty to start but files are named accordingly, user sees the behavior, cool. and then they start fleshing them out. 

vs, they've defined 3, but they just don't do anything. no error, just noop.

Reply:
That's already how it works. If there's no changes we just don't have to make a copy.